File: plugins/shipyard/bin/hook-runner.py
#!/usr/bin/env python3
"""Cross-platform hook launcher for Shipyard.

Eliminates bash-specific syntax from hooks.json by handling:
- Plugin root discovery (from __file__, no shell variable expansion needed)
- SHIPYARD_DATA resolution (calls shipyard-data internally)
- CWD management (changes to project dir)
- Script dispatch (runs target hook script)

Usage in hooks.json:
  "command": "python3 \"${CLAUDE_PLUGIN_ROOT}/bin/hook-runner.py\" <script-name>"

Falls back to discovering plugin root from its own file path if
CLAUDE_PLUGIN_ROOT is not set (Windows compatibility).
"""

import importlib.util
import os
import sys


def discover_plugin_root():
    """Find the plugin root directory.

    Primary: derive from this file's location (bin/hook-runner.py → plugin root).
    Fallback: CLAUDE_PLUGIN_ROOT environment variable.
    """
    # This file lives in plugins/shipyard/bin/
    bin_dir = os.path.dirname(os.path.abspath(__file__))
    plugin_root = os.path.dirname(bin_dir)

    # Sanity check: plugin.json should exist
    plugin_json = os.path.join(plugin_root, '.claude-plugin', 'plugin.json')
    if os.path.isfile(plugin_json):
        return plugin_root

    # Fallback to env var — but only if it actually contains plugin.json
    # (prevents env-based redirect to an arbitrary directory)
    env_root = os.environ.get('CLAUDE_PLUGIN_ROOT', '')
    if env_root and os.path.isfile(os.path.join(env_root, '.claude-plugin', 'plugin.json')):
        return env_root

    # Hard fail: refuse to run with an undetermined plugin root
    print(
        f'hook-runner: cannot locate plugin root. Tried {plugin_root} and '
        f'CLAUDE_PLUGIN_ROOT={env_root!r}. plugin.json not found in either.',
        file=sys.stderr
    )
    sys.exit(1)


def resolve_shipyard_data(plugin_root):
    """Compute SHIPYARD_DATA, preferring the env var set by hook-runner.mjs.

    Production fast path: hook-runner.mjs imports the Node resolver in-process
    and passes SHIPYARD_DATA via env, so we read it here and return immediately.
    No subprocess spawn — saves ~80ms per Edit hook (R7).

    Slow path (tests, direct invocation): subprocess to bin/shipyard-resolver.mjs.
    The Node resolver remains the single source of truth so this stays in sync
    with shipyard-data.mjs and shipyard-context.mjs.

    Worktree behavior: the resolver returns the PARENT repo's data dir even
    when called from inside a worktree, so all worktrees of one project share
    state. See DECISIONS.md D1.
    """
    env_value = os.environ.get('SHIPYARD_DATA', '')
    if env_value:
        return env_value

    import subprocess
    resolver = os.path.join(plugin_root, 'bin', 'shipyard-resolver.mjs')
    try:
        # 2s timeout (R8): git rev-parse is bounded; long hangs are real
        # problems we should fail fast on, not paper over with a 10s wait.
        result = subprocess.run(
            ['node', resolver, 'data-dir'],
            capture_output=True, text=True, timeout=2,
        )
        if result.returncode == 0:
            return result.stdout.strip()
    except (FileNotFoundError, subprocess.TimeoutExpired):
        pass
    # Hard fail rather than fall back to a divergent path. The auto-approve
    # hook depends on this matching the value the skill computes; a silent
    # fallback would split state between two directories.
    print(
        'hook-runner: shipyard-resolver.mjs failed. Ensure Node 18+ is on PATH.',
        file=sys.stderr,
    )
    sys.exit(1)


def run_script(script_path):
    """Run a Python script by loading it as a module and calling main()."""
    spec = importlib.util.spec_from_file_location('hook_script', script_path)
    if spec is None or spec.loader is None:
        print(f'hook-runner: cannot load {script_path}', file=sys.stderr)
        sys.exit(1)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    if hasattr(module, 'main'):
        module.main()


# Map short names to script paths (relative to project-files/scripts/)
SCRIPT_MAP = {
    'auto-approve-data': 'auto-approve-data.py',
    'tdd-check': 'tdd-check.py',
    'session-guard': 'session-guard.py',
    'worktree-branch': 'worktree-branch.py',
    'loop-detect': 'loop-detect.py',
    'on-commit': 'on-commit.py',
    'cwd-restore': 'cwd-restore.py',
    'post-compact': 'post-compact.py',
}


def main():
    if len(sys.argv) < 2:
        print('Usage: hook-runner.py <script-name>', file=sys.stderr)
        sys.exit(1)

    script_name = sys.argv[1]

    if script_name not in SCRIPT_MAP:
        print(f'hook-runner: unknown script "{script_name}"', file=sys.stderr)
        print(f'Available: {", ".join(sorted(SCRIPT_MAP))}', file=sys.stderr)
        sys.exit(1)

    plugin_root = discover_plugin_root()
    shipyard_data = resolve_shipyard_data(plugin_root)

    # Set environment for the target script. Always overwrite CLAUDE_PLUGIN_ROOT
    # with the discovered value so it matches the value discover_plugin_root()
    # already preferred (file-derived path beats inherited env). Using
    # setdefault here would let a stale env value win — confusing.
    os.environ['SHIPYARD_DATA'] = shipyard_data
    os.environ['CLAUDE_PLUGIN_ROOT'] = plugin_root

    # Change to project directory if available
    project_dir = os.environ.get('CLAUDE_PROJECT_DIR', '')
    if project_dir and os.path.isdir(project_dir):
        try:
            os.chdir(project_dir)
        except OSError:
            pass

    # Resolve and run the target script
    script_file = SCRIPT_MAP[script_name]
    script_path = os.path.join(plugin_root, 'project-files', 'scripts', script_file)

    if not os.path.isfile(script_path):
        print(f'hook-runner: script not found: {script_path}', file=sys.stderr)
        sys.exit(1)

    run_script(script_path)


if __name__ == '__main__':
    main()

File: plugins/shipyard/bin/hook-runner.mjs
#!/usr/bin/env node
/**
 * Cross-platform hook dispatcher for Shipyard.
 *
 * In-process Node.js implementation. Dispatches hook events directly to
 * Node modules under `bin/hooks/` via `import()`. The previous version
 * spawned a Python child process — required Python 3 on every user
 * machine and added ~80ms of process startup to every Edit/Write hook.
 *
 * Phase H4 has shipped: the Python implementation is gone. The runner
 * is purely in-process.
 *
 * Usage: node hook-runner.mjs <hook-name>
 * Stdin: hook JSON input piped through to the dispatched module.
 */

import { readFileSync, existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { getDataDir } from "./shipyard-resolver.mjs";

const __dirname = dirname(fileURLToPath(import.meta.url));
const HOOK_NAME_RE = /^[a-z][a-z0-9-]{0,40}$/;
const HOOKS_DIR = join(__dirname, "hooks");

const scriptName = process.argv[2];
if (!scriptName) {
  process.stderr.write("hook-runner: missing script name\n");
  process.exit(1);
}

// Defense in depth: validate the hook name against a strict allowlist
// before constructing a module path. Even though the name comes from
// hooks.json (which we trust), a misconfigured matcher or future env-var
// interpolation could feed in something unexpected.
if (!HOOK_NAME_RE.test(scriptName)) {
  process.stderr.write(`hook-runner: invalid hook name ${JSON.stringify(scriptName)}\n`);
  process.exit(1);
}

// Read stdin (hook JSON) before doing anything else. The Python path
// piped stdin into the Python child; we now parse it inline.
let stdin = "";
try {
  stdin = readFileSync(0, "utf8");
} catch {
  // No stdin is fine for some hooks (e.g., post-compact)
}

// Compute SHIPYARD_DATA in-process and pass it to the hook via env.
// In-process import is free relative to a process spawn — the resolver's
// worktree-parent detection runs once per hook fire either way.
//
// Customer regression note: this used to be silently swallowed when
// getDataDir threw, masking the root cause for months. The stderr line
// is how customers will know which case they're in.
const env = { ...process.env };
if (!env.SHIPYARD_DATA) {
  try {
    env.SHIPYARD_DATA = getDataDir({ silent: true });
  } catch (err) {
    const errName = err?.name || "Error";
    const firstLine = (err?.message || String(err)).split("\n")[0].slice(0, 240);
    process.stderr.write(
      `shipyard hook-runner: getDataDir threw (${errName}): ${firstLine}\n`,
    );
  }
}

// Parse stdin JSON into hookInput. Hooks that need raw text get an empty
// object — they should treat absent fields as "no work to do" and exit 0.
let hookInput = {};
if (stdin) {
  try {
    hookInput = JSON.parse(stdin);
  } catch {
    hookInput = {};
  }
}

// In-process dispatch via dynamic import.
const modulePath = join(HOOKS_DIR, `${scriptName}.mjs`);
if (!existsSync(modulePath)) {
  process.stderr.write(`hook-runner: no such hook ${JSON.stringify(scriptName)}\n`);
  process.exit(1);
}
// Mirror env for the child via process.env (the in-process hook reads
// process.env directly). Restore on exit.
const origEnv = {};
for (const k of Object.keys(env)) {
  origEnv[k] = process.env[k];
  process.env[k] = env[k];
}
try {
  const url = "file://" + modulePath;
  const mod = await import(url);
  if (typeof mod.run !== "function") {
    process.stderr.write(
      `hook-runner: hook ${scriptName} does not export run()\n`,
    );
    process.exit(1);
  }
  const code = await mod.run(hookInput, env);
  process.exit(typeof code === "number" ? code : 0);
} catch (err) {
  process.stderr.write(`hook-runner: hook ${scriptName} threw: ${err?.message || err}\n`);
  process.exit(1);
} finally {
  // Restore env on exit (though we exit immediately afterward).
  for (const k of Object.keys(origEnv)) {
    if (origEnv[k] === undefined) delete process.env[k];
    else process.env[k] = origEnv[k];
  }
}
